Add zoom with new "zoom" privilege

bendeutsch opened the pull request:
Rebased version of #1953 by @ShadowNinja , which is itself a rebased version of #583 by @EXio4 .

Many critics of the original version did not like zooming being available to everyone by default. A Lua API was proposed to allow games / mods to affect the zoom. This version simply requires a new privilege "zoom", not granted to singleplayer by default, to use the zoom.

Comments welcome.

est31 commented on src/guiKeyChangeMenu.cpp:
```diff
@@ -414,5 +415,6 @@ void GUIKeyChangeMenu::init_keys()
 	this->add_key(GUI_ID_KEY_NOCLIP_BUTTON,    wgettext("Toggle noclip"),    "keymap_noclip");
 	this->add_key(GUI_ID_KEY_RANGE_BUTTON,     wgettext("Range select"),     "keymap_rangeselect");
 	this->add_key(GUI_ID_KEY_DUMP_BUTTON,      wgettext("Print stacks"),     "keymap_print_debug_stacks");
+	this->add_key(GUI_ID_KEY_ZOOM_BUTTON,      wgettext("Zoom"),          "keymap_zoom");
```
Can you align it with the other lines?

est31 commented:
Also don't forget documenting your settings changes in the builtin/settingstypes.txt file.

est31 commented:
If those two are done: :+1:

est31 commented:
great feature, like it.

bendeutsch commented:
Indented the line and added to `builtin/settingtypes.txt`. I noted that the default `zoom_fov` of "15" is currently outside the range of `fov` itself (30 to 160) – any thoughts on this?

The feature itself is not my idea, by the way, it's been around for a while it seems :stuck_out_tongue:

paramat commented:
I assume mods can set privs, for example to grant this only when a special item is crafted?

est31 commented:
@paramat yes.

est31 commented:
@bendeutsch  The fov setting is for the default field of view, so thats okay.

Maybe squash your last two commits?

bendeutsch commented:
@paramat while this could be used to achieve an in-game effect, I'd recommend someone™ write an actual Lua API for the FOV instead. That could be used not only for wielding telescopes but also for a visual effect during sprinting or a fish-eye curse.

@est31 The second commit actually belongs more to the first commit, but then I'd attribute authorship of my settings description to EXio4, who may not want it. The third commit is semantically separate from the first two since it adds the privilege requirement to something that was originally complete and functional.

est31 commented:
We can't get it super perfect, if that's the case we'd get nothing changed. In the super perfect case, this value would be modified by client side lua and the engine would never have to take care of sending this info over the wire.

SmallJoker commented:
Here's the addition to the Lua API: 38471a6

bendeutsch commented:
@SmallJoker : I'd reverse the order of checks to ask for the privilege first. I see it more as a tool, like `fly`, `/tp`, or `/kick` (also keeping in mind that server admins get all privileges by default).

Also, what about players who play with a non-standard FOV? I think mods should either *also* be able to "offset" the FOV (think sprinting in Minecraft), or (my favorite) the Lua API should assume the standard FOV, and the actual FOV should be some combination of the standard FOV, the player-set FOV, and the Lua-set FOV.

I'd like to add a transition animation as well. Maybe the Lua API needs a field for "transition speed in degrees per second" as well.

Lastly, setting the FOV via Lua is one aspect; another one is changing the behavior of the zoom key via mods. An example of the first would be using a telescope item, an example of the second is bionic eye implants. A visual effect would be neat, too, say in the form of a full-screen texture overlay. This sounds like it's outside the scope of this pull request, but if we keep the zoom key as an in-game effect, the client needs to know what to do without server intervention.

The network packet keeps getting bigger. Is there any reason not to skip sending individual integers, and go to a fully serialized key-value map?

SmallJoker commented:
The idea behind that commit was to allow shooter mods to switch the FOV quicky when the weapon was activated or selected. Giving the zoom privilege the priority would disallow mods to overwrite the player FOV. I agree with the offset idea but it could be a better to allow to overwrite it like the physics - using factors to modify the player's default setting.
[New commit to be added here soon]

bendeutsch commented:
Giving the zoom privilege the priority still requires the zoom key to be pushed, though, so I don't think it's too much of a problem.

Would a factor instead of an offset cover all use cases? Might as well enumerate them: sprinting effect, scopes (tele- and sniper), and… what else?

bendeutsch commented:
Incidently, the code is a mess at the moment :-/ Both the privilege check and the (new) transition animation take place in the camera's `update` method. The transition at least should move to the `step` function, but the calculation of the "desired fov" requires the both player controls and privilege checks at the moment. Could use some refactoring.

everamzah commented:
Oh for the love of Pete.

paramat commented:
I support the original PR but oppose a transition animation, it's unnecessary, lets keep this simple.
Use of a telescope type item or gunsight is a sudden change in FOV anyway.
Changes in FOV are disorientating and nauseating when there is a smooth transition, this is something i notice watching Minecraft videos.

est31 commented on src/camera.cpp:
```diff
@@ -387,14 +389,27 @@ void Camera::update(LocalPlayer* player, f32 frametime, f32 busytime,
 	fov_degrees = MYMAX(fov_degrees, 10.0);
 	fov_degrees = MYMIN(fov_degrees, 170.0);
+	f32 fov_step_size = 500.0 * frametime; // 500 degrees per second
+	if ( fov_degrees < m_fov_degrees - fov_step_size) {
```
no space between ( and fov_degrees

est31 commented:
:-1: to transitional animation as well.

bendeutsch commented:
Alright, we're making progress! I've removed the transition animation (and rebased to master in the process). I consider this finished, now.

The Lua API should probably be a new pull request.

paramat commented:
One last thing, please document in minetest.conf? although this can be auto-generated this is not done often, and a new setting should be present immediately.

bendeutsch commented:
Added documentation to minetest.conf.example; I assume this is what you meant?

paramat commented:
Yeah sorry that's what i meant.

nerzhul commented:
:+1:

est31 commented:
2c31b79235dd83de753fce5890c5797e149048b8
d1df09841d0eac7a88f638676b80ec848522cca5
82494b8cc199fb92248c6c30ce7ef18b7e7133ae
